File: problems/CSES/Dynamic_Programming/book_shop/book_shop.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, money = map(int, input().split())
    prices = list(map(int, input().split()))
    pages = list(map(int, input().split()))

    dp = [0] * (money + 1)

    for i in range(0, n):
        curr_price = prices[i]
        curr_pages = pages[i]

        for x in range(money, curr_price - 1, -1):
            dp[x] = max(dp[x],
                        dp[x - curr_price] + curr_pages)

    print(dp[money])

main()

File: problems/CSES/Dynamic_Programming/book_shop/book_shop.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, money;
    cin >> n >> money;

    vector<int> prices(n);
    vector<int> pages(n);

    for (int& price : prices) {
        cin >> price;
    }

    for (int& page : pages) {
        cin >> page;
    }

    vector<int> dp(money + 1, 0);

    for (int i = 0; i < n; ++i) {
        int currPrice = prices[i];
        int currPages = pages[i];

        for (int x = money; x >= currPrice; --x) {
            dp[x] = max(
                dp[x],
                dp[x - currPrice] + currPages
            );
        }
    }

    cout << dp[money] << '\n';

    return 0;
}
